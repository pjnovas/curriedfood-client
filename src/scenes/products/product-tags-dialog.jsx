import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet } from 'react-native';
import { Portal, Dialog, Button } from 'react-native-paper';
import { get, noop } from 'lodash';

import Theme from 'theme';
import { composeHooks } from 'utils/language';
import Chips from 'components/chips';

export const ProductTagsDialog = ({
  setTagValue,
  onDismiss,
  onSubmitPress,
  title,
  allTags,
  tags
}) => (
  <Portal>
    <Dialog style={styles.layout} visible dismissable onDismiss={onDismiss}>
      <Dialog.Title style={styles.title}>{title}</Dialog.Title>
      <Dialog.Content style={styles.content}>
        <Chips list={allTags} value={tags} onChange={setTagValue} />
      </Dialog.Content>
      <Dialog.Actions style={styles.actions}>
        <Button onPress={onDismiss}>Cancelar</Button>
        <Button
          dark
          mode="contained"
          style={styles.okButton}
          onPress={onSubmitPress}
        >
          OK
        </Button>
      </Dialog.Actions>
    </Dialog>
  </Portal>
);

ProductTagsDialog.propTypes = {
  onChange: PropTypes.func,
  onDismiss: PropTypes.func,
  onSubmitPress: PropTypes.func,
  title: PropTypes.string,
  allTags: PropTypes.arrayOf(PropTypes.string),
  tags: PropTypes.string
};

ProductTagsDialog.defaultProps = {
  onChange: noop,
  onDismiss: noop,
  onSubmitPress: noop,
  title: '',
  allTags: [],
  tags: ''
};

const styles = StyleSheet.create({
  layout: {},
  title: {
    color: Theme.colors.text,
    textAlign: 'center'
  },
  content: {
    minHeight: 100,
    maxHeight: 400
  },
  actions: {},
  okButton: {
    marginHorizontal: 10
  }
});

export const useProduct = ({ product, onSubmit, tagField }) => {
  // const allTags = []; // TODO: get tags for a product by fetching all products
  const [tags, setTagValue] = useState(get(product, tagField, ''));
  // TODO: Save product

  return {
    tags,
    title: product.name,
    setTagValue,
    onSubmitPress: () => {
      console.log(tags);
      onSubmit(tags);
    }
  };
};

export default composeHooks({ useProduct })(ProductTagsDialog);
