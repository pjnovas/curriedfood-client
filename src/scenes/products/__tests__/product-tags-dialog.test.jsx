import * as React from 'react';
import { noop } from 'lodash';
import { ProductTagsDialog, useProduct } from '../product-tags-dialog';

const allTags = ['verduleria', 'super', 'carniceria'];

const props = {
  title: 'Aceitunas',
  allTags,
  tags: 'super,carniceria',
  onSubmitPress: noop,
  setTagValue: noop,
  onDismiss: noop
};

takeSnapshots(
  {
    default: {},
    withData: props
  },
  ProductTagsDialog
);

test('useProduct', () => {
  const product = {
    shop_tags: 'super,carniceria',
    name: 'Aceitunas'
  };

  const onSubmit = jest.fn();
  const tagField = 'shop_tags';

  const spyUseState = jest.spyOn(React, 'useState');
  React.useState.mockImplementation(() => [product.shop_tags, jest.fn]);

  const result = useProduct({ product, onSubmit, tagField });

  expect(React.useState).toBeCalledWith(product.shop_tags);
  expect(result.tags).toEqual(product.shop_tags);
  expect(result.title).toEqual(product.name);

  result.onSubmitPress();
  expect(onSubmit).toBeCalledWith(result.tags);

  spyUseState.mockReset();
});
