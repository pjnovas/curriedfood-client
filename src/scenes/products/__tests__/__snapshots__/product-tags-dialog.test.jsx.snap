// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
<withTheme(Portal)>
  <withTheme(Dialog)
    dismissable={true}
    onDismiss={[Function]}
    style={Object {}}
    visible={true}
  >
    <withTheme(Dialog.Title)
      style={
        Object {
          "color": "#ffffff",
          "textAlign": "center",
        }
      }
    />
    <Dialog.Content
      style={
        Object {
          "maxHeight": 400,
          "minHeight": 100,
        }
      }
    >
      <Component
        list={Array []}
        value=""
      />
    </Dialog.Content>
    <Dialog.Actions
      style={Object {}}
    >
      <withTheme(Button)
        onPress={[Function]}
      >
        Cancelar
      </withTheme(Button)>
      <withTheme(Button)
        dark={true}
        mode="contained"
        onPress={[Function]}
        style={
          Object {
            "marginHorizontal": 10,
          }
        }
      >
        OK
      </withTheme(Button)>
    </Dialog.Actions>
  </withTheme(Dialog)>
</withTheme(Portal)>
`;

exports[`withData 1`] = `
<withTheme(Portal)>
  <withTheme(Dialog)
    dismissable={true}
    onDismiss={[Function]}
    style={Object {}}
    visible={true}
  >
    <withTheme(Dialog.Title)
      style={
        Object {
          "color": "#ffffff",
          "textAlign": "center",
        }
      }
    >
      Aceitunas
    </withTheme(Dialog.Title)>
    <Dialog.Content
      style={
        Object {
          "maxHeight": 400,
          "minHeight": 100,
        }
      }
    >
      <Component
        list={
          Array [
            "verduleria",
            "super",
            "carniceria",
          ]
        }
        onChange={[Function]}
        value="super,carniceria"
      />
    </Dialog.Content>
    <Dialog.Actions
      style={Object {}}
    >
      <withTheme(Button)
        onPress={[Function]}
      >
        Cancelar
      </withTheme(Button)>
      <withTheme(Button)
        dark={true}
        mode="contained"
        onPress={[Function]}
        style={
          Object {
            "marginHorizontal": 10,
          }
        }
      >
        OK
      </withTheme(Button)>
    </Dialog.Actions>
  </withTheme(Dialog)>
</withTheme(Portal)>
`;
